---
title: "Homework 4"
author: "Abhijit Dasgupta"
output:
  word_document: default
  html_document: default
---

```{r setup, include=F}
knitr::opts_chunk$set(include=FALSE, message=F, warning=F, cache=TRUE)
```


In this homework we will play with some publicly available RNA-Seq data. 
Information about the experiment is available [here](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE123519). Briefly,
this is a mouse experiment where the authors were interested in PIK3CA mutation
variants. This interest is derived from PIK3CA's implication in glioblastoma tumorogenesis. There are 26 mice with different engineered genetics. 

The data and experimental design are available as a zip file and excel file respectively in Teams Week 4 files (GSE123519_RAW.zip and GSE123519design.xlsx). 

### Prelim work
1. Download and unzip the data file in your project data directory. You should get 26 text files, perhaps in a sub-directory. Note this, since you'll need it
in the next step. Also download the experimental design file.

### Homework

#### Data preparation



1. Read all 26 files into a list in R, where the list consists of 26 data frames
    ```{r}
    library(here)
    library(rio)
    dat <- import_list(dir(here('data/GSE123519_RAW/'), full.names = TRUE))
    ```
1. For each of the data frames perform the following data processing
    a. Keep only the variables `tracking_id`, `gene_id`, and `locus` through `FPKM`.
    b. Change the gene names in `gene_id` to capital letters (use `stringr::str_to_upper` or other equivalent functions)
    c. Split the `locus` variable into a `chromosome` and a `location` variable.
    d. Filter **out** all observations with FPKM less than 1 (no real scientific reason, just because)
    ```{r}
    library(tidyverse)
    dat1 <- dat %>% 
        map(select, tracking_id, gene_id, locus:FPKM) %>% 
        map(mutate, gene_id = stringr::str_to_upper(gene_id)) %>% 
        map(separate, locus, c('chromosome','location'), sep = ':') %>% 
        map(filter, FPKM >= 1)
    ```
1. Now make a single data set stacking the processed datasets, adding a column that
specifies the sample identifier of each observation. You might like to call this column `sampleID` to make the next step simpler
    ```{r}
    dat2 <- bind_rows(dat1, .id = 'sampleID') %>% 
        mutate(sampleID = str_extract(sampleID, 'GSM\\d+'))
    ```
1. Import the design information into a data frame. Add the design information to 
the genetic dataset you created above, ensuring that ids and genotype
specification (stored in the column `Class`) are aligned.
    ```{r}
    design <- import(here('data/GSE123519design.xlsx'))
    dat2 <- dat2 %>% left_join(design)
    ```
1. Note that there is an add-on to the genotype information, in the form `-1`,
`-2` to specify which particular sample it is. This is unnecessary and would
create problems for any comparative analyses using t-tests or ANOVA later.
Please remove it so that you just have genotype information.
    ```{r}
    dat2 <- dat2 %>% 
        mutate(Class = str_remove(Class, '-[0-9]'))
    ```

1. There is one idiosyncratic genotype specification which is a re-rerun. Fix this to 
match the other genotypes. 
    ```{r}
    dat2 <- dat2 %>% 
        mutate(Class = str_remove(Class,'-re-run-59'))
    ```
    
1. Clean the `chromosome` values so that you only have data of the form "chr##" where ## refers to the chromosome number.
    ```{r}
    dat2 <- dat2 %>% 
        mutate(chromosome = str_extract(chromosome, 'chr[[:alnum:]]+'))
    
    ```

1. Save the dataset using `saveRDS` to a .rds file
    ```{r}
    saveRDS(dat2, 'gene_data.rds', compress = T)
    ```


You can choose whether you fix all the genotype information before or after you join the design dataset to the genetic dataset. Whatever makes sense to you.

#### Data analysis

1. Provide the following tables:
    - median FPKM levels by Class
    - How many unique genes are interrogated per chromosome
    ```{r}
    dat2 <- readRDS('gene_data.rds')
    dat2 %>% group_by(Class) %>% summarise(FPKM = median(FPKM, na.rm=T))
    dat2 %>% group_by(chromosome) %>% summarize(N = length(unique(gene_id)))
    ```
    
1. How many unique genes of the form PIK3C* are present in the dataset
```{r week4-1}
dat2 %>% filter(str_detect(gene_id, 'PIK3C')) %>%
    summarize(length(unique(gene_id)))
```

1. What are the relative frequencies of each PIK3CA variant among the 26 mice. Assume a missing value means the variant is not present.
```{r week4-2}
dat2 %>% filter(str_detect(gene_id, 'PIK3C')) %>%
    select(sampleID, gene_id) %>% 
    distinct() %>% count(gene_id) %>% mutate(rel_freq = n/26)

```

1. Perform a statistical test to see if FPKM expression levels are different at PIK3C* genes compared to all other genes. (the correct test is actually complicated, but doing t-test, wilcoxon test or permutation tests are all allowed here)
```{r week4-3}
dat2 %>% mutate(pik3c_ind = ifelse(str_detect(gene_id, 'PIK3C'), 1, 0)) %>% 
    t.test(FPKM ~ pik3c_ind, data = .) %>% 
    broom::tidy()
```

1. Draw a barplot showing the median FPKM levels by Class
```{r week4-4}
dat2 %>% group_by(Class) %>% summarize(FPKM = median(FPKM, na.rm=T)) %>% 
    ggplot(aes(Class, FPKM))+geom_bar(stat='identity') + coord_flip()

# or 

dat2 %>% ggplot(aes(Class, FPKM)) + stat_summary(geom='bar', fun = median) +
    coord_flip()
```


    
