<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Practical R</title>
    <meta charset="utf-8" />
    <meta name="author" content="Abhijit Dasgupta" />
    <script src="00_intro_files/header-attrs-2.6.6/header-attrs.js"></script>
    <link href="00_intro_files/tachyons-4.12.0/tachyons.min.css" rel="stylesheet" />
    <link href="00_intro_files/panelset-0.2.3.9000/panelset.css" rel="stylesheet" />
    <script src="00_intro_files/panelset-0.2.3.9000/panelset.js"></script>
    <link href="00_intro_files/xaringanExtra-extra-styles-0.2.3.9000/xaringanExtra-extra-styles.css" rel="stylesheet" />
    <link href="00_intro_files/countdown-0.3.5/countdown.css" rel="stylesheet" />
    <script src="00_intro_files/countdown-0.3.5/countdown.js"></script>
    <link rel="stylesheet" href="../css/xaringan-themer.css" type="text/css" />
    <link rel="stylesheet" href="../css/custom.css" type="text/css" />
    <link rel="stylesheet" href="../css/sfah.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Practical R
### Abhijit Dasgupta
### 25 January, 2021

---










---
class: middle,center,inverse

# About this class

---

## Learning Objectives

- Run R and RStudio, making use of inherent R features
- Find and make use of the extensive packages (R add-ons) available for analyzing biological and other forms of data
- Load, manipulate, and combine data to make it amenable to further analyses
- Visualize data with extensive graphics capabilities of R (including ggplot)
- Use R to run statistical models and hypothesis tests and report results conforming to standards expected in scientific journals
- Write reports using the powerful `rmarkdown` package and its derivatives

---

## Plan


```
NULL
```


---

## Teaching materials

1. The primary materials for this class will be available at https://www.araastat.com/AZ_PracticalR (Link in Teams)
1. Tutorials from RStudio's [Primers](https://rstudio.cloud/learn/primers) will be assigned for practice
1. The [Resources](../../resources.html) page has various online free resources, including several books and cheatsheets. Some readings may be assigned from these resources

---

## Communication

+ Primarily via Teams using the class channel
+ Questions and comments for each week should be kept within that week's chat 
+ If you have questions/comments, please post on Teams rather than sending me an e-mail or a direct message. This is so we can all participate, help each other, and benefit from all the questions and answers

---
class: middle,center,inverse

# Why use R?

---

class: middle, center

&lt;img src="../img/R.jpg" height="561" width="724"/&gt;

---

# What does R look like?

.left-column30[
+ A .heatinline[scripting] language
  - Provide instructions to the computer
  - in a .saltinline[structured] manner
  - to do statistical analysis
]
.right-column30[


```r
# Determining frequencies of breast cancer subtypes

type_frequencies &lt;- 
  breast_cancer %&gt;% 
    mutate(luminalA = ifelse(ER == '+' &amp; PR == '+' &amp; HER2 == '-', 1, 0),
         luminalB = ifelse(ER == '+' &amp; PR == '-' &amp; HER2 == '+', 1, 0),
         her2 = ifelse(ER == '-' &amp; PR == '-' &amp; HER2 == '+', 1, 0),
         basal = ifelse(ER == '-' &amp; PR == '-' &amp; HER2 == '-', 1, 0)) %&gt;% 
    mutate(type = case_when(
      luminalA == 1 ~ "Luminal A",
      luminalB == 1 ~ "Luminal B",
      her2 == 1 ~ "HER2",
      basal == 1 ~ "Basal",
      TRUE ~ NA)) %&gt;%
    count(type)
```
]

---

# What does R look like?

.left-column30[
+ Start with a data set
]
.right-column30[


```r
# Determining frequencies of breast cancer subtypes

type_frequencies &lt;- 
* breast_cancer %&gt;%
    mutate(luminalA = ifelse(ER == '+' &amp; PR == '+' &amp; HER2 == '-', 1, 0),
         luminalB = ifelse(ER == '+' &amp; PR == '-' &amp; HER2 == '+', 1, 0),
         her2 = ifelse(ER == '-' &amp; PR == '-' &amp; HER2 == '+', 1, 0),
         basal = ifelse(ER == '-' &amp; PR == '-' &amp; HER2 == '-', 1, 0)) %&gt;% 
    mutate(type = case_when(
      luminalA == 1 ~ "Luminal A",
      luminalB == 1 ~ "Luminal B",
      her2 == 1 ~ "HER2",
      basal == 1 ~ "Basal",
      TRUE ~ NA)) %&gt;%
    count(type)
```
]

---

# What does R look like?

.left-column30[
+ Start with a data set
+ Create new variables from old variables
]
.right-column30[


```r
# Determining frequencies of breast cancer subtypes

type_frequencies &lt;- 
  breast_cancer %&gt;% 
*   mutate(luminalA = ifelse(ER == '+' &amp; PR == '+' &amp; HER2 == '-', 1, 0),
*        luminalB = ifelse(ER == '+' &amp; PR == '-' &amp; HER2 == '+', 1, 0),
*        her2 = ifelse(ER == '-' &amp; PR == '-' &amp; HER2 == '+', 1, 0),
*        basal = ifelse(ER == '-' &amp; PR == '-' &amp; HER2 == '-', 1, 0)) %&gt;%
    mutate(type = case_when(
      luminalA == 1 ~ "Luminal A",
      luminalB == 1 ~ "Luminal B",
      her2 == 1 ~ "HER2",
      basal == 1 ~ "Basal",
      TRUE ~ NA)) %&gt;%
    count(type)
```
]

---

# What does R look like?

.left-column30[
+ Start with a data set
+ Create new variables from old variables
+ Deal with missing values
]
.right-column30[


```r
# Determining frequencies of breast cancer subtypes

type_frequencies &lt;- 
  breast_cancer %&gt;% 
    mutate(luminalA = ifelse(ER == '+' &amp; PR == '+' &amp; HER2 == '-', 1, 0), 
         luminalB = ifelse(ER == '+' &amp; PR == '-' &amp; HER2 == '+', 1, 0), 
         her2 = ifelse(ER == '-' &amp; PR == '-' &amp; HER2 == '+', 1, 0), 
         basal = ifelse(ER == '-' &amp; PR == '-' &amp; HER2 == '-', 1, 0)) %&gt;% 
    mutate(type = case_when(
      luminalA == 1 ~ "Luminal A",
      luminalB == 1 ~ "Luminal B",
      her2 == 1 ~ "HER2",
      basal == 1 ~ "Basal",
*     TRUE ~ NA)) %&gt;%
    count(type)
```
]

---

# What does R look like?

.left-column30[
+ Start with a data set
+ Create new variables from old variables
+ Deal with missing values
+ Find the frequencies
]
.right-column30[


```r
# Determining frequencies of breast cancer subtypes

type_frequencies &lt;- 
  breast_cancer %&gt;% 
    mutate(luminalA = ifelse(ER == '+' &amp; PR == '+' &amp; HER2 == '-', 1, 0), 
         luminalB = ifelse(ER == '+' &amp; PR == '-' &amp; HER2 == '+', 1, 0), 
         her2 = ifelse(ER == '-' &amp; PR == '-' &amp; HER2 == '+', 1, 0), 
         basal = ifelse(ER == '-' &amp; PR == '-' &amp; HER2 == '-', 1, 0)) %&gt;% 
    mutate(type = case_when(
      luminalA == 1 ~ "Luminal A",
      luminalB == 1 ~ "Luminal B",
      her2 == 1 ~ "HER2",
      basal == 1 ~ "Basal",
      TRUE ~ NA)) %&gt;%
*   count(type)
```
]

---

# What does R look like?

.left-column30[
+ Start with a data set
+ Create new variables from old variables
+ Deal with missing values
+ Find the frequencies
+ Save the results in an **object**
]
.right-column30[


```r
# Determining frequencies of breast cancer subtypes

*type_frequencies &lt;-
  breast_cancer %&gt;% 
    mutate(luminalA = ifelse(ER == '+' &amp; PR == '+' &amp; HER2 == '-', 1, 0), 
         luminalB = ifelse(ER == '+' &amp; PR == '-' &amp; HER2 == '+', 1, 0), 
         her2 = ifelse(ER == '-' &amp; PR == '-' &amp; HER2 == '+', 1, 0), 
         basal = ifelse(ER == '-' &amp; PR == '-' &amp; HER2 == '-', 1, 0)) %&gt;% 
    mutate(type = case_when(
      luminalA == 1 ~ "Luminal A",
      luminalB == 1 ~ "Luminal B",
      her2 == 1 ~ "HER2",
      basal == 1 ~ "Basal",
      TRUE ~ NA)) %&gt;%
    count(type) 
```
]

---

# What does R look like?

.left-column30[
+ Start with a data set
+ Create new variables from old variables
+ Deal with missing values
+ Find the frequencies
+ Save the results in an **object**
+ .heatinline[Comment on what you're doing]
]
.right-column30[


```r
*# Determining frequencies of breast cancer subtypes

type_frequencies &lt;- 
  breast_cancer %&gt;% 
    mutate(luminalA = ifelse(ER == '+' &amp; PR == '+' &amp; HER2 == '-', 1, 0), 
         luminalB = ifelse(ER == '+' &amp; PR == '-' &amp; HER2 == '+', 1, 0), 
         her2 = ifelse(ER == '-' &amp; PR == '-' &amp; HER2 == '+', 1, 0), 
         basal = ifelse(ER == '-' &amp; PR == '-' &amp; HER2 == '-', 1, 0)) %&gt;% 
    mutate(type = case_when(
      luminalA == 1 ~ "Luminal A",
      luminalB == 1 ~ "Luminal B",
      her2 == 1 ~ "HER2",
      basal == 1 ~ "Basal",
      TRUE ~ NA)) %&gt;%
    count(type) 
```
]

--

.center[This is an example of a .heatinline[pipeline] in R. We'll develop different aspects of this
progressively throughout the course]

---
class: middle,center,inverse

# But, I already know Excel/Graphpad/Prism/SPSS!!


---
# Why use a scripting language for analysis?

.pull-left[
.large[Pros:]

1. .heatinline[Have to think]
1. Reproducible (custom) workflows
1. Much less error-prone
1. Much lower costs to repeat analyses, or as you learn more
1. Easily leverage work of smarter developers
1. Easier to work with larger datasets (more than size of screen)
]
.pull-right[
.large[Cons:]

1. Have to type
1. Have to know the language
1. Higher initial startup cost
1. .heatinline[Have to think]
]

---

# Why use a scripting language for analysis?

.pull-left[
.large[Pros:]

1. **Have to think**
1. Reproducible (custom) workflows
1. Much less error-prone
1. Much lower costs to repeat analyses, or as you learn more
1. Easily leverage work of smarter developers
1. Easier to work with larger datasets (more than size of screen)
]
.pull-right[
+ You're giving instructions to a fast but stupid machine
+ This machine will do .heatinline[**exactly**] what you tell it
+ The machine is capable of amazing things
+ Can't just _menu-mine_ and try things that __seem__ to be what you want

#### With great power comes great responsibility

#### But also great benefits

]

---

# Why use a scripting language for analysis?

.pull-left[
.large[Pros:]

1. Have to think
1. Reproducible (custom) workflows
1. **Much less error-prone**
1. Much lower costs to repeat analyses, or as you learn more
1. Easily leverage work of smarter developers
1. Easier to work with larger datasets (more than size of screen)
]
.pull-right[
+ If your code is not right, it won't run
  - Can be frustrating
  - But if it runs you're much more confident
+ If you screw up in Excel
  - almost impossible to recover
+ You have much more control over what you're doing
]

---

# Why use a scripting language for analysis?

.pull-left[
.large[Pros:]

1. Have to think
1. Reproducible (custom) workflows
1. Much less error-prone
1. Much lower costs to repeat analyses, or as you learn more
1. **Easily leverage work of smarter developers**
1. Easier to work with larger datasets (more than size of screen)
]
.pull-right[
+ Can use _modules_ or _packages_ developed by others
  - `tidyverse`, `Seurat`, `ggplot2`
+ Can "steal" code from others (provided license allows)
]

---

# Why use a scripting language for analysis?

.pull-left[
.large[Pros:]

1. Have to think
1. Reproducible (custom) workflows
1. Much less error-prone
1. Much lower costs to repeat analyses, or as you learn more
1. Easily leverage work of smarter developers
1. **Easier to work with larger datasets (more than size of screen)**
]
.pull-right[
Good luck working with

+ GWAS data
+ fMRI data
+ Stocks and bonds data
+ Sports data
+ Many more ...

in a unified environment  

Learn once, use everywhere
]

---

# Why use &lt;svg style="height:0.8em;top:.04em;position:relative;fill:steelblue;" viewBox="0 0 581 512"&gt;&lt;path d="M581 226.6C581 119.1 450.9 32 290.5 32S0 119.1 0 226.6C0 322.4 103.3 402 239.4 418.1V480h99.1v-61.5c24.3-2.7 47.6-7.4 69.4-13.9L448 480h112l-67.4-113.7c54.5-35.4 88.4-84.9 88.4-139.7zm-466.8 14.5c0-73.5 98.9-133 220.8-133s211.9 40.7 211.9 133c0 50.1-26.5 85-70.3 106.4-2.4-1.6-4.7-2.9-6.4-3.7-10.2-5.2-27.8-10.5-27.8-10.5s86.6-6.4 86.6-92.7-90.6-87.9-90.6-87.9h-199V361c-74.1-21.5-125.2-67.1-125.2-119.9zm225.1 38.3v-55.6c57.8 0 87.8-6.8 87.8 27.3 0 36.5-38.2 28.3-87.8 28.3zm-.9 72.5H365c10.8 0 18.9 11.7 24 19.2-16.1 1.9-33 2.8-50.6 2.9v-22.1z"/&gt;&lt;/svg&gt;?

+ Specializes in statistics and data visualization
+ Flexible
  - ~~_If you can do it_~~ __How you can do it__
+ Large ecosystem
  - Over 16,000 packages, 1500+ dedicated to bioinformatics
  - Can read from most sources of data
  - Generic and specialized analyses
+ Fantastic community
  - Twitter, StackOverflow, blogosphere, conferences, online books

---

# Why use &lt;svg style="height:0.8em;top:.04em;position:relative;fill:steelblue;" viewBox="0 0 581 512"&gt;&lt;path d="M581 226.6C581 119.1 450.9 32 290.5 32S0 119.1 0 226.6C0 322.4 103.3 402 239.4 418.1V480h99.1v-61.5c24.3-2.7 47.6-7.4 69.4-13.9L448 480h112l-67.4-113.7c54.5-35.4 88.4-84.9 88.4-139.7zm-466.8 14.5c0-73.5 98.9-133 220.8-133s211.9 40.7 211.9 133c0 50.1-26.5 85-70.3 106.4-2.4-1.6-4.7-2.9-6.4-3.7-10.2-5.2-27.8-10.5-27.8-10.5s86.6-6.4 86.6-92.7-90.6-87.9-90.6-87.9h-199V361c-74.1-21.5-125.2-67.1-125.2-119.9zm225.1 38.3v-55.6c57.8 0 87.8-6.8 87.8 27.3 0 36.5-38.2 28.3-87.8 28.3zm-.9 72.5H365c10.8 0 18.9 11.7 24 19.2-16.1 1.9-33 2.8-50.6 2.9v-22.1z"/&gt;&lt;/svg&gt;?

R is a very high-quality product that is accepted and used widely in 
government agencies, corporations and universities worldwide

- Standard data analytic software in bioinformatics, behavioral health and many aspects of quantitative finance
- Increasingly used in pharma, economics, political science and engineering

--

R is open-source, in that anyone can see the actual code and validate the 
computations directly

---

# Why use &lt;svg style="height:0.8em;top:.04em;position:relative;fill:steelblue;" viewBox="0 0 581 512"&gt;&lt;path d="M581 226.6C581 119.1 450.9 32 290.5 32S0 119.1 0 226.6C0 322.4 103.3 402 239.4 418.1V480h99.1v-61.5c24.3-2.7 47.6-7.4 69.4-13.9L448 480h112l-67.4-113.7c54.5-35.4 88.4-84.9 88.4-139.7zm-466.8 14.5c0-73.5 98.9-133 220.8-133s211.9 40.7 211.9 133c0 50.1-26.5 85-70.3 106.4-2.4-1.6-4.7-2.9-6.4-3.7-10.2-5.2-27.8-10.5-27.8-10.5s86.6-6.4 86.6-92.7-90.6-87.9-90.6-87.9h-199V361c-74.1-21.5-125.2-67.1-125.2-119.9zm225.1 38.3v-55.6c57.8 0 87.8-6.8 87.8 27.3 0 36.5-38.2 28.3-87.8 28.3zm-.9 72.5H365c10.8 0 18.9 11.7 24 19.2-16.1 1.9-33 2.8-50.6 2.9v-22.1z"/&gt;&lt;/svg&gt;?

- R has a well-deserved reputation for being a great data visualization 
tool, with users being able to create complex, customizable graphs with relative 
ease
- As a scripting language, it allows the same workflows to be coded and re-used.
- You can set up workflows to validate data, in terms of data quality and missingness, 
which avoids visual inspection which can be time-consuming and mistake-prone. 
- R can handle large datasets, and can work with multiple datasets at the same time

---

# Why use &lt;svg style="height:0.8em;top:.04em;position:relative;fill:steelblue;" viewBox="0 0 581 512"&gt;&lt;path d="M581 226.6C581 119.1 450.9 32 290.5 32S0 119.1 0 226.6C0 322.4 103.3 402 239.4 418.1V480h99.1v-61.5c24.3-2.7 47.6-7.4 69.4-13.9L448 480h112l-67.4-113.7c54.5-35.4 88.4-84.9 88.4-139.7zm-466.8 14.5c0-73.5 98.9-133 220.8-133s211.9 40.7 211.9 133c0 50.1-26.5 85-70.3 106.4-2.4-1.6-4.7-2.9-6.4-3.7-10.2-5.2-27.8-10.5-27.8-10.5s86.6-6.4 86.6-92.7-90.6-87.9-90.6-87.9h-199V361c-74.1-21.5-125.2-67.1-125.2-119.9zm225.1 38.3v-55.6c57.8 0 87.8-6.8 87.8 27.3 0 36.5-38.2 28.3-87.8 28.3zm-.9 72.5H365c10.8 0 18.9 11.7 24 19.2-16.1 1.9-33 2.8-50.6 2.9v-22.1z"/&gt;&lt;/svg&gt;?

Specialized packages available for many domains

- Bioinformatics
- Econometrics
- Maps and spatial analytics 
- Text mining and Natural Language Processing

The [CRAN Task Views](https://cran.r-project.org/web/views/) 
provide curated lists of packages based on different domains

The [Bioconductor Project](https://www.bioconductor.org) is THE source for bioinformatics packages in R. It is the gold standard for many bioinformatic workflows



---

# Things &lt;svg style="height:0.8em;top:.04em;position:relative;fill:steelblue;" viewBox="0 0 581 512"&gt;&lt;path d="M581 226.6C581 119.1 450.9 32 290.5 32S0 119.1 0 226.6C0 322.4 103.3 402 239.4 418.1V480h99.1v-61.5c24.3-2.7 47.6-7.4 69.4-13.9L448 480h112l-67.4-113.7c54.5-35.4 88.4-84.9 88.4-139.7zm-466.8 14.5c0-73.5 98.9-133 220.8-133s211.9 40.7 211.9 133c0 50.1-26.5 85-70.3 106.4-2.4-1.6-4.7-2.9-6.4-3.7-10.2-5.2-27.8-10.5-27.8-10.5s86.6-6.4 86.6-92.7-90.6-87.9-90.6-87.9h-199V361c-74.1-21.5-125.2-67.1-125.2-119.9zm225.1 38.3v-55.6c57.8 0 87.8-6.8 87.8 27.3 0 36.5-38.2 28.3-87.8 28.3zm-.9 72.5H365c10.8 0 18.9 11.7 24 19.2-16.1 1.9-33 2.8-50.6 2.9v-22.1z"/&gt;&lt;/svg&gt; is not (in this class)

--

- The 18th letter of the English alphabet

--

- A magic incantation that will produce an analysis

--

- Just something the cool kids are doing

--

- Point-and-click, automatic, WYSIWYG (What you see is what you get)
    - So it's not Excel, SPSS, Prism, GraphPad
    - It's much more!!!

---

# A note on Excel &lt;svg style="height:0.8em;top:.04em;position:relative;fill:steelblue;" viewBox="0 0 384 512"&gt;&lt;path d="M224 136V0H24C10.7 0 0 10.7 0 24v464c0 13.3 10.7 24 24 24h336c13.3 0 24-10.7 24-24V160H248c-13.2 0-24-10.8-24-24zm60.1 106.5L224 336l60.1 93.5c5.1 8-.6 18.5-10.1 18.5h-34.9c-4.4 0-8.5-2.4-10.6-6.3C208.9 405.5 192 373 192 373c-6.4 14.8-10 20-36.6 68.8-2.1 3.9-6.1 6.3-10.5 6.3H110c-9.5 0-15.2-10.5-10.1-18.5l60.3-93.5-60.3-93.5c-5.2-8 .6-18.5 10.1-18.5h34.8c4.4 0 8.5 2.4 10.6 6.3 26.1 48.8 20 33.6 36.6 68.5 0 0 6.1-11.7 36.6-68.5 2.1-3.9 6.2-6.3 10.6-6.3H274c9.5-.1 15.2 10.4 10.1 18.4zM384 121.9v6.1H256V0h6.1c6.4 0 12.5 2.5 17 7l97.9 98c4.5 4.5 7 10.6 7 16.9z"/&gt;&lt;/svg&gt;

.pull-left[
   

Excel is omnipresent, so it becomes the default data medium

It is great for many things, including _quick-and-dirty_ analyses

.heatinline[It can be error-prone]

It needs to be backed up

It has size limitations

Takes a lot of effort to do complex analyses
]

--

.pull-right[
+ No way to reproduce analyses without macros
+ No way to document what you are doing
+ Excel has some nasty default behavior
  - Guess what the `MAR1` gene gets recorded as?
+ Very hard to recover from errors
  - Shift of one error (off by one row or column)
  - Google _Duke Potti_ or _Reinhart Rogoff Herndon_
]

---
class: middle,center,inverse

# Using RStudio and organizing projects

---

## RStudio

#### I'll be doing a demo here, but document the main points in the following slides

---

## Open a new script 

.pull-left[
&lt;img src="../img/NewScript.png" width="4139" /&gt;
]

.pull-right[
You will start by opening a new script

- This is a blank canvas
- You will write R code (instructions) in this script
- This is just a text editor, so you can copy, paste, edit just as in Notepad
- Good practice is to write code in the script window, then **send** the code to R
]


---

## Run some code

.pull-left[

Write some code, then send it to the console, where R is running

- You can use the button, but
- I prefer you learn the keyboard shortcut
- Ctrl/Cmd + Enter sends either
  - current line under cursor
  - highlighted code
]
.pull-right[
![](../img/RunCode.png)]

---
class: middle,center,inverse

# RStudio Projects

---

# RStudio Projects

&lt;img src="../img/project1.png" width="4139" /&gt;

---

## RStudio Projects
&lt;img src="../img/project2.png" width="4139" /&gt;

---

## RStudio Projects
&lt;img src="../img/project3.png" width="4139" /&gt;

---

## RStudio Projects
&lt;img src="../img/project4.png" width="4139" /&gt;

---

## RStudio Projects
&lt;img src="../img/project5.png" width="4139" /&gt;

---

## RStudio Projects
&lt;img src="../img/project6.png" width="4139" /&gt;

---

## RStudio Projects
&lt;img src="../img/project7.png" width="4139" /&gt;

---
name: proj1

## RStudio Projects

When you create a Project, the following obvious things happen:

1. RStudio puts you into the right directory/folder
2. Creates a .Rproj file containing project options
    + You can double-click on the .Rproj file to open the project in RStudio
3. Displays the project name in the project toolbar (right top of the window)

---

## RStudio Projects

The following not-so-obvious things happen:

1. A new R session (process) is started
2. The .Rprofile file in the project’s main directory (if any) is sourced by R
3. The .RData file in the project’s main directory is loaded (this can be controlled by an option).
4. The .Rhistory file in the project’s main directory is loaded into the RStudio History pane (and used for Console Up/Down arrow command history).
5. The current working directory is set to the project directory.
6. Previously edited source documents are restored into editor tabs, and
7. Other RStudio settings (e.g. active tabs, splitter positions, etc.) are restored to where they were the last time the project was closed.

---

## RStudio Projects

I use [Projects]() so that:

1. I'm always in the right directory for the project
2. I don't contaminate one project's analysis with another (different sandboxes)
3. I can access different projects quickly
4. I can version control them (Git) easily (topic for beyond this class)
5. I can customize options per project

.footnote[A great reference for git from a R user's perspective is Jenny Bryan's excellent [happygitwithr.com](https://happygitwithr.com/)]
---

## RStudio Projects
&lt;img src="../img/project8.png" width="4139" /&gt;

---
class: middle,center,inverse

# Taking notes and putting together reports: R Markdown

---








## R Markdown

.left-column30[
+ There are some choices you might need to make
  + Document is like a paper
  + Presentation is like PowerPoint
  + Shiny is an interactive web app developed in R. May talk about it towards the end
  + Various packages also provide templates for RMarkdown documents

]
.right-column30[
![](../img/rmd2.png)
]

---



.left-column30[
+ Elements on the right are output formats
  - Documents produce Word, PDF or HTML
  - Presentations produce PowerPoint, PDF, or HTML
]
.right-column30[
![](../img/rmd2.png)
]

---

```
---
title: "Untitled"
author: "Abhijit Dasgupta"
date: "9/11/2019"
output: html_document
---
```

This determines the title and author, and the output type. Typically don't modify the 
entry in `output` for now

--

````markdown
```{r cars}
summary(cars)
```
````

This is a code chunk. RMarkdown evaluates this chunk of code and replaces the code 
with the results. This code chunk happens to be named "cars". The naming is optional but useful.

---

.pull-left[
### RMarkdown

````markdown
```{r cars}
summary(cars)
```
````
]
.pull-right[
### Results


```r
summary(cars)
```

```
     speed           dist       
 Min.   : 4.0   Min.   :  2.00  
 1st Qu.:12.0   1st Qu.: 26.00  
 Median :15.0   Median : 36.00  
 Mean   :15.4   Mean   : 42.98  
 3rd Qu.:19.0   3rd Qu.: 56.00  
 Max.   :25.0   Max.   :120.00  
```
]

---

.pull-left[
### RMarkdown

````markdown
```{r}
library(tableone) # Use a package
kableone(CreateTableOne(data=airquality), 
         format='html')
```
````
]
.pull-right[
### Results


```r
library(tableone) # Use a package
kableone(CreateTableOne(data=airquality), 
         format='html')
```

&lt;table&gt;
 &lt;thead&gt;
  &lt;tr&gt;
   &lt;th style="text-align:left;"&gt;   &lt;/th&gt;
   &lt;th style="text-align:left;"&gt; Overall &lt;/th&gt;
  &lt;/tr&gt;
 &lt;/thead&gt;
&lt;tbody&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; n &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 153 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; Ozone (mean (SD)) &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 42.13 (32.99) &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; Solar.R (mean (SD)) &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 185.93 (90.06) &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; Wind (mean (SD)) &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 9.96 (3.52) &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; Temp (mean (SD)) &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 77.88 (9.47) &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; Month (mean (SD)) &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 6.99 (1.42) &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; Day (mean (SD)) &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 15.80 (8.86) &lt;/td&gt;
  &lt;/tr&gt;
&lt;/tbody&gt;
&lt;/table&gt;

]

&lt;br&gt;
&lt;br&gt;
The code chunk on the left gets **replaced** by the table on the right in your document


---

![](../img/rmd5.png)

---

## Inline code evaluation

.pull-left[

### RMarkdown
The airquality data set has   
`` `r
nrow(airquality)` `` observations


The average ozone level is `` `r
mean(airquality$Ozone)` `` parts per billion
]
.pull-right[

### Results

The airquality data set has 153 observations

The average ozone level is   
42.1293103 `` parts per billion

]

---
class: middle,center,inverse

# Data structures in R

---































# Objects in R

.pull-left[
Let's start with the `airquality` data.

- It is an object
- of class `class(airquality)` = data.frame


How about each column? 

Let's look at the Ozone and Wind columns

- We can access them using `airquality$Ozone` and `airquality$Wind`
  - `class(airquality$Ozone)` = integer
  - `class(airquality$Wind)` = numeric

]
.pull-right[

```
   Ozone Solar.R Wind Temp Month Day
1     41     190  7.4   67     5   1
2     36     118  8.0   72     5   2
3     12     149 12.6   74     5   3
4     18     313 11.5   62     5   4
5     NA      NA 14.3   56     5   5
6     28      NA 14.9   66     5   6
7     23     299  8.6   65     5   7
8     19      99 13.8   59     5   8
9      8      19 20.1   61     5   9
10    NA     194  8.6   69     5  10
11     7      NA  6.9   74     5  11
12    16     256  9.7   69     5  12
13    11     290  9.2   66     5  13
14    14     274 10.9   68     5  14
15    18      65 13.2   58     5  15
16    14     334 11.5   64     5  16
17    34     307 12.0   66     5  17
18     6      78 18.4   57     5  18
19    30     322 11.5   68     5  19
20    11      44  9.7   62     5  20
21     1       8  9.7   59     5  21
22    11     320 16.6   73     5  22
23     4      25  9.7   61     5  23
24    32      92 12.0   61     5  24
25    NA      66 16.6   57     5  25
```

]

---

# Objects in R


```r
head(iris)
```

```
  Sepal.Length Sepal.Width Petal.Length Petal.Width Species
1          5.1         3.5          1.4         0.2  setosa
2          4.9         3.0          1.4         0.2  setosa
3          4.7         3.2          1.3         0.2  setosa
4          4.6         3.1          1.5         0.2  setosa
5          5.0         3.6          1.4         0.2  setosa
6          5.4         3.9          1.7         0.4  setosa
```

```r
str(iris)
```

```
'data.frame':	150 obs. of  5 variables:
 $ Sepal.Length: num  5.1 4.9 4.7 4.6 5 5.4 4.6 5 4.4 4.9 ...
 $ Sepal.Width : num  3.5 3 3.2 3.1 3.6 3.9 3.4 3.4 2.9 3.1 ...
 $ Petal.Length: num  1.4 1.4 1.3 1.5 1.4 1.7 1.4 1.5 1.4 1.5 ...
 $ Petal.Width : num  0.2 0.2 0.2 0.2 0.2 0.4 0.3 0.2 0.2 0.1 ...
 $ Species     : Factor w/ 3 levels "setosa","versicolor",..: 1 1 1 1 1 1 1 1 1 1 ...
```

Now we see another type of object, a `factor`

---

# Objects in R


```r
library(ggplot2)
str(midwest)
```

```
tibble [437 × 28] (S3: tbl_df/tbl/data.frame)
 $ PID                 : int [1:437] 561 562 563 564 565 566 567 568 569 570 ...
 $ county              : chr [1:437] "ADAMS" "ALEXANDER" "BOND" "BOONE" ...
 $ state               : chr [1:437] "IL" "IL" "IL" "IL" ...
 $ area                : num [1:437] 0.052 0.014 0.022 0.017 0.018 0.05 0.017 0.027 0.024 0.058 ...
 $ poptotal            : int [1:437] 66090 10626 14991 30806 5836 35688 5322 16805 13437 173025 ...
 $ popdensity          : num [1:437] 1271 759 681 1812 324 ...
 $ popwhite            : int [1:437] 63917 7054 14477 29344 5264 35157 5298 16519 13384 146506 ...
 $ popblack            : int [1:437] 1702 3496 429 127 547 50 1 111 16 16559 ...
 $ popamerindian       : int [1:437] 98 19 35 46 14 65 8 30 8 331 ...
 $ popasian            : int [1:437] 249 48 16 150 5 195 15 61 23 8033 ...
 $ popother            : int [1:437] 124 9 34 1139 6 221 0 84 6 1596 ...
 $ percwhite           : num [1:437] 96.7 66.4 96.6 95.3 90.2 ...
 $ percblack           : num [1:437] 2.575 32.9 2.862 0.412 9.373 ...
 $ percamerindan       : num [1:437] 0.148 0.179 0.233 0.149 0.24 ...
 $ percasian           : num [1:437] 0.3768 0.4517 0.1067 0.4869 0.0857 ...
 $ percother           : num [1:437] 0.1876 0.0847 0.2268 3.6973 0.1028 ...
 $ popadults           : int [1:437] 43298 6724 9669 19272 3979 23444 3583 11323 8825 95971 ...
 $ perchsd             : num [1:437] 75.1 59.7 69.3 75.5 68.9 ...
 $ percollege          : num [1:437] 19.6 11.2 17 17.3 14.5 ...
 $ percprof            : num [1:437] 4.36 2.87 4.49 4.2 3.37 ...
 $ poppovertyknown     : int [1:437] 63628 10529 14235 30337 4815 35107 5241 16455 13081 154934 ...
 $ percpovertyknown    : num [1:437] 96.3 99.1 95 98.5 82.5 ...
 $ percbelowpoverty    : num [1:437] 13.15 32.24 12.07 7.21 13.52 ...
 $ percchildbelowpovert: num [1:437] 18 45.8 14 11.2 13 ...
 $ percadultpoverty    : num [1:437] 11.01 27.39 10.85 5.54 11.14 ...
 $ percelderlypoverty  : num [1:437] 12.44 25.23 12.7 6.22 19.2 ...
 $ inmetro             : int [1:437] 0 0 0 1 0 0 0 0 0 1 ...
 $ category            : chr [1:437] "AAR" "LHR" "AAR" "ALU" ...
```
 Here we have a `character`.
 
---

# Objects in R

The most common types of data we see are `numeric`, `character`, `factor`. You can also see `Date` and `logical`

You can test to see if data is of a particular type, or convert from one data type to another

&lt;table&gt;
 &lt;thead&gt;
  &lt;tr&gt;
   &lt;th style="text-align:left;"&gt; Data type &lt;/th&gt;
   &lt;th style="text-align:left;"&gt; Test &lt;/th&gt;
   &lt;th style="text-align:left;"&gt; Convert &lt;/th&gt;
  &lt;/tr&gt;
 &lt;/thead&gt;
&lt;tbody&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; numeric &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; is.numeric &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; as.numeric &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; character &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; is.character &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; as.character &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; factor &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; is.factor &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; as.factor &lt;/td&gt;
  &lt;/tr&gt;
&lt;/tbody&gt;
&lt;/table&gt;

--

This conversion is important. Why?

<div class="countdown" id="timer_6020f5d8" style="right:0;bottom:0;" data-warnwhen="0">
<code class="countdown-time"><span class="countdown-digits minutes">05</span><span class="countdown-digits colon">:</span><span class="countdown-digits seconds">00</span></code>
</div>

---

# Factors

Factors are uniquely an R thing. 

They are meant to represent categorical data (gender, race, state, phenotype)

They look like character vectors, but internally act like integers, so you have to 
be a bit careful with them

--

Whenever you're in doubt, convert them to characters using `as.character`.

We'll see the utility of factors when we do data munging, summaries and modeling




---

## Every object in R has a name

You give an object a name using the syntax `name &lt;- object`

Naming conventions: 

1. Snake_case or pothole_case
1. CamelCase
1. Some.people.use.periods

I'm partial to `snake_case`.

The point here is to make expressive names using English so you know what 
is stored in the name.

---

## A silly exercise

From the iris dataset, save each column into a new object, giving it a name. Then
see what kind of data that object contains.

<div class="countdown" id="timer_6020f62d" style="right:0;bottom:0;" data-warnwhen="0">
<code class="countdown-time"><span class="countdown-digits minutes">05</span><span class="countdown-digits colon">:</span><span class="countdown-digits seconds">00</span></code>
</div>

---

## Bigger objects

### Scalar -&gt; vector (array) -&gt; matrix (2-d array)

--

+ A scalar is a single number or word
+ A vector is a bunch of scalars arranged in a row or column
+ A matrix is a bunch of scalars arranged in rows and columns

#### Each of these **must** be of the same data type

---

## Examples


```r
2
```

```
[1] 2
```

--


```r
c(2,3,4,5,6)
```

```
[1] 2 3 4 5 6
```

&gt; `c()` is the concatenate function

--


```r
matrix(c(1,2,3,4), byrow = T, nrow = 2)
```

```
     [,1] [,2]
[1,]    1    2
[2,]    3    4
```

---
class:middle,center

# Data comes in many flavors

---

## Heterogeneous data

From Excel, we are familiar with keeping different kinds of data together in a spreadsheet

+ Expression levels (numeric)
+ Gene names (character)
+ Date of experiment (Date)


In R, the objects that can hold heterogeneous data are `data.frame` and `list`

---
class:middle, center

# Data sets

---

## Typical data structure

+ Data is typically in a rectangular format
  
    + spreadsheet, database table
    + CSV (comma-separated values) or TSV (tab-separated values) files

+ Characteristic
  
    + Rows are observations
    + Columns are variables
    + Each column has the same number of observations

&gt; [__Tidy data__](https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html) is a particularly amenable format for data analysis. 
 
---

# The `data.frame`

Dataframes are the primary mode of storing datasets in R

They were revolutionary in that they kept heterogeneous data together

They share properties of both a __matrix__ and a __list__


```r
class(airquality)
```

```
[1] "data.frame"
```

&gt; Technically, a data.frame is a list of vectors (or objects, generally) of the same length  

---
class: middle, center

# Load some data

---

We'll load the `spine` dataset into R.

To do this, download the data from Teams, and store it in the _data_ folder in your
project.

Then, in the Environment pane, import it using the **Import Dataset** button. You will 
use the `From text (readr)` option




---

![](../img/readr1.png)

---

![](../img/readr2.png)

---

![](../img/readr3.png)

---
class:center, middle

# A digression: Lists and Matrices

---




# Matrices

A __matrix__ is a rectangular array of data _of the same type_


```r
matrix(0, nrow=2, ncol=4)
```

```
     [,1] [,2] [,3] [,4]
[1,]    0    0    0    0
[2,]    0    0    0    0
```

```r
matrix(letters, nrow=2)
```

```
     [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13]
[1,] "a"  "c"  "e"  "g"  "i"  "k"  "m"  "o"  "q"  "s"   "u"   "w"   "y"  
[2,] "b"  "d"  "f"  "h"  "j"  "l"  "n"  "p"  "r"  "t"   "v"   "x"   "z"  
```

```r
matrix(letters, nrow=2, byrow=T)
```

```
     [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13]
[1,] "a"  "b"  "c"  "d"  "e"  "f"  "g"  "h"  "i"  "j"   "k"   "l"   "m"  
[2,] "n"  "o"  "p"  "q"  "r"  "s"  "t"  "u"  "v"  "w"   "x"   "y"   "z"  
```

---

# Matrices

You can create a matrix from a set of _vectors_ of the same length

```r
x &lt;- c(1,2,3,4)
y &lt;- c(10,20,30,40)
```
Put columns together

```r
cbind(c(1,2,3,4), c(10,20,30,40)) ## Column bind
```

```
     [,1] [,2]
[1,]    1   10
[2,]    2   20
[3,]    3   30
[4,]    4   40
```

---

# Matrices

You can create a matrix from a set of _vectors_ of the same length

```r
x &lt;- c(1,2,3,4)
y &lt;- c(10,20,30,40)
```
Put rows together

```r
example_matrix &lt;- rbind(c(1,2,3,4), c(10,20,30,40)) ## Row bind
example_matrix
```

```
     [,1] [,2] [,3] [,4]
[1,]    1    2    3    4
[2,]   10   20   30   40
```

---

# Extracting elements


```r
example_matrix 
```

```
     [,1] [,2] [,3] [,4]
[1,]    1    2    3    4
[2,]   10   20   30   40
```

```r
example_matrix[1,] ## Extracts 1st row
```

```
[1] 1 2 3 4
```

```r
example_matrix[,2:3] ## extracts 2nd &amp; 3rd columns
```

```
     [,1] [,2]
[1,]    2    3
[2,]   20   30
```

```r
example_matrix[1,4]
```

```
[1] 4
```

---

# Matrix properties

```r
example_matrix
```

```
     [,1] [,2] [,3] [,4]
[1,]    1    2    3    4
[2,]   10   20   30   40
```

```r
nrow(example_matrix) ## Number of rows
```

```
[1] 2
```

```r
ncol(example_matrix) ## Number of columns
```

```
[1] 4
```

```r
dim(example_matrix) ## shortcut for above
```

```
[1] 2 4
```

---

# Matrix arithmetic

```r
example_matrix
```

```
     [,1] [,2] [,3] [,4]
[1,]    1    2    3    4
[2,]   10   20   30   40
```

```r
example_matrix + 5 ## Add 5 to each element
```

```
     [,1] [,2] [,3] [,4]
[1,]    6    7    8    9
[2,]   15   25   35   45
```

```r
example_matrix * 2 ## Multiply each element by 2
```

```
     [,1] [,2] [,3] [,4]
[1,]    2    4    6    8
[2,]   20   40   60   80
```

---

# Two matrices

```r
example_matrix
```

```
     [,1] [,2] [,3] [,4]
[1,]    1    2    3    4
[2,]   10   20   30   40
```

```r
example_matrix2 &lt;- rbind(3:6, 9:12)
example_matrix2
```

```
     [,1] [,2] [,3] [,4]
[1,]    3    4    5    6
[2,]    9   10   11   12
```

```r
example_matrix + example_matrix2
```

```
     [,1] [,2] [,3] [,4]
[1,]    4    6    8   10
[2,]   19   30   41   52
```

---

# Two matrices

```r
example_matrix
```

```
     [,1] [,2] [,3] [,4]
[1,]    1    2    3    4
[2,]   10   20   30   40
```

```r
example_matrix2
```

```
     [,1] [,2] [,3] [,4]
[1,]    3    4    5    6
[2,]    9   10   11   12
```

```r
example_matrix * example_matrix2 ## Not matrix multiplication, but element-wise multiplication
```

```
     [,1] [,2] [,3] [,4]
[1,]    3    8   15   24
[2,]   90  200  330  480
```

---

# Two matrices

```r
rbind(example_matrix, example_matrix2)
```

```
     [,1] [,2] [,3] [,4]
[1,]    1    2    3    4
[2,]   10   20   30   40
[3,]    3    4    5    6
[4,]    9   10   11   12
```

```r
cbind(example_matrix, example_matrix2)
```

```
     [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8]
[1,]    1    2    3    4    3    4    5    6
[2,]   10   20   30   40    9   10   11   12
```

---

# Two matrices

```r
dim(example_matrix2)
```

```
[1] 2 4
```

```r
t(example_matrix2) ## Transpose of a matrix
```

```
     [,1] [,2]
[1,]    3    9
[2,]    4   10
[3,]    5   11
[4,]    6   12
```

```r
example_matrix %*% t(example_matrix2) ## Matrix multiplication
```

```
     [,1] [,2]
[1,]   50  110
[2,]  500 1100
```

---

# Lists

Lists are collections of arbitrary objects in R

```r
example_list &lt;- list(c('Andy','Brian','Harry'), 
                     c(12, 16, 16), 
                     c(TRUE, TRUE, FALSE), 
                     matrix(1, nrow=2, ncol=3))
example_list
```

```
[[1]]
[1] "Andy"  "Brian" "Harry"

[[2]]
[1] 12 16 16

[[3]]
[1]  TRUE  TRUE FALSE

[[4]]
     [,1] [,2] [,3]
[1,]    1    1    1
[2,]    1    1    1
```

---

# Extracting elements from lists

```r
example_list[[3]]
```

```
[1]  TRUE  TRUE FALSE
```

```r
example_list[1:2]
```

```
[[1]]
[1] "Andy"  "Brian" "Harry"

[[2]]
[1] 12 16 16
```

---

# Extracting elements from lists

```r
example_list[[4]]
```

```
     [,1] [,2] [,3]
[1,]    1    1    1
[2,]    1    1    1
```

```r
class(example_list[[4]])
```

```
[1] "matrix" "array" 
```

```r
example_list[[4]][1,]
```

```
[1] 1 1 1
```

---

# Named lists

```r
example_named_list &lt;- list('Names' = c('Andy','Brian','Harry'), 
                     "YearsOfEducation" = c(12, 16, 16), 
                     "Married" = c(TRUE, TRUE, FALSE), 
                     'something' = matrix(1, nrow=2, ncol=3))

example_named_list[['Names']]
```

```
[1] "Andy"  "Brian" "Harry"
```

```r
example_named_list$Names
```

```
[1] "Andy"  "Brian" "Harry"
```

```r
example_named_list$Names[3]
```

```
[1] "Harry"
```

---
class: middle, center

# Back to a Data Frame

---

# Data frames

A data.frame object is a __named list__ where each element is of the same length

You can use both _matrix_ and _list_ functions to operate on data.frame objects!!

---

# Data Frames


```r
head(data_spine)
```

```
  Pelvic.incidence Pelvic.tilt Lumbar.lordosis.angle Sacral.slope Pelvic.radius
1         63.02782   22.552586              39.60912     40.47523      98.67292
2         39.05695   10.060991              25.01538     28.99596     114.40543
3         68.83202   22.218482              50.09219     46.61354     105.98514
4         69.29701   24.652878              44.31124     44.64413     101.86850
5         49.71286    9.652075              28.31741     40.06078     108.16872
6         40.25020   13.921907              25.12495     26.32829     130.32787
  Degree.spondylolisthesis Pelvic.slope Direct.tilt Thoracic.slope
1                -0.254400    0.7445035     12.5661        14.5386
2                 4.564259    0.4151857     12.8874        17.5323
3                -3.530317    0.4748892     26.8343        17.4861
4                11.211523    0.3693453     23.5603        12.7074
5                 7.918501    0.5433605     35.4940        15.9546
6                 2.230652    0.7899929     29.3230        12.0036
  Cervical.tilt Sacrum.angle Scoliosis.slope Class.attribute
1      15.30468   -28.658501         43.5123        Abnormal
2      16.78486   -25.530607         16.1102        Abnormal
3      16.65897   -29.031888         19.2221        Abnormal
4      11.42447   -30.470246         18.8329        Abnormal
5       8.87237   -16.378376         24.9171        Abnormal
6      10.40462    -1.512209          9.6548        Abnormal
```

---

# Data Frames

```r
dim(data_spine)
```

```
[1] 310  13
```

```r
nrow(data_spine)
```

```
[1] 310
```

```r
data_spine_small &lt;- data_spine[1:4,] ## Matrix operation
```

---

# Data Frames

```r
data_spine_small[,2] ## Matrix extraction by position
```

```
[1] 22.55259 10.06099 22.21848 24.65288
```

```r
data_spine_small[[2]] ## List extraction by position
```

```
[1] 22.55259 10.06099 22.21848 24.65288
```

---

# Data Frames

```r
data_spine_small[['Pelvic.tilt']] ## Named list extraction
```

```
[1] 22.55259 10.06099 22.21848 24.65288
```

```r
data_spine_small[,'Pelvic.tilt'] ## Data frame named column extraction
```

```
[1] 22.55259 10.06099 22.21848 24.65288
```

```r
data_spine_small$Pelvic.tilt ## Dollar sign extraction
```

```
[1] 22.55259 10.06099 22.21848 24.65288
```

---

# Data Frames

My preference is for

1. _data frame named column extraction_ `data_spine_small[,'Pelvic.tilt']`, 
2. _named list extraction_ `data_spine_small[['Pelvic.tilt']]`
3. _Dollar-based extraction_ `data_spine_small$Pelvic.tilt`


---

# Data Frames

```r
names(data_spine_small)
```

```
 [1] "Pelvic.incidence"         "Pelvic.tilt"             
 [3] "Lumbar.lordosis.angle"    "Sacral.slope"            
 [5] "Pelvic.radius"            "Degree.spondylolisthesis"
 [7] "Pelvic.slope"             "Direct.tilt"             
 [9] "Thoracic.slope"           "Cervical.tilt"           
[11] "Sacrum.angle"             "Scoliosis.slope"         
[13] "Class.attribute"         
```

```r
data_spine_small[,c('Pelvic.tilt', 'Pelvic.slope','Class.attribute')]
```

```
  Pelvic.tilt Pelvic.slope Class.attribute
1    22.55259    0.7445035        Abnormal
2    10.06099    0.4151857        Abnormal
3    22.21848    0.4748892        Abnormal
4    24.65288    0.3693453        Abnormal
```


---
class: middle,center,inverse

# Functions

---



---
class: middle,center,inverse

# Naming things

---











## Why do we need names?

Everything in R is an object that can be named

Unless you name an object, it doesn't stay in the R environment

You name an object using the assignment operator `&lt;-`  
I read this as a left arrow

For example

```r
spine_data &lt;- read.csv('data/Dataset_spine.csv')
```

Here, we read the CSV file into R, and then save it in the current R session by naming it `spine_data`

---

## Syntactic names

Following these rules will always give you valid object names without the need for additional trouble.

1. Contains letters, numbers, `.` or `_` characters
1. Starts with a letter or `.` not followed by a number
1. Is not a reserved word 

--

So `.2way` is not a syntactically valid name

---

## Going wild

You can actually use **any** name for an object, as long as you encase it in backticks

The name can then include spaces, special characters, start with a number, etc.

Still, don't use reserved words.

--

.pull-left[

You can do 

`` `First variable` ``

`` `2 many bikes` ``

]
.pull-right[
.heat[But should you?!!]
]

---

## Naming conventions

There are several prevalent naming conventions in R

+ snake case, e.g., `my_variable`
+ lower camel case, e.g., `myVariable`
+ upper camel case, e.g., `MyVariable`
+ period separated, e.g., `my.variable`
+ lower case, e.g., `myvariable`

Choose one and stick with it. It will save you some sanity.


-----

&gt; The **janitor** package has a nice function called `clean_names` which will 
transform the column names in a data.frame to follow a particular naming convention, 
with the default being snake case.

---
background-image: url(../img/janitor_clean_names.png)
background-size: contain

---

## Resources

The following commands, typed in the R console, will get you to the corresponding help files.
+ `?make.names`
+ `?assign`

---
class: middle,center,inverse

# Brackets

---

layout: true


## Brackets

There are three kinds of brackets in R

---


.fatinline[`[]`] are used for extracting elements from arrays, matrices, data frames.

.panelset[
.panel[.panel-name[Arrays]
- `x[3]` is the 3rd element of an array `x`
- `d[1,3]` is the element in the 1st row and 3rd column of a matrix/data frame `d`
- `d[2,]` is the entire first __row__ of a matrix/data frame `d`
- `d[,4]` is the entire 4th __column__ of a matrix/data frame `d`
- `d[,'gender']` is the column named _gender_ in a _data frame_ `d`
]
.panel[.panel-name[Lists]

To extract single elements, you use a .acidinline[double bracket:]
- `lst[[3]]` is the third element of the list `lst`
- `lst[['apple']]` is the element named _apple_ in a named list `lst`

To extract a sub-list, you use .acidinline[single brackets]:
+ `lst[1:3]` creates a sub-list with the first 3 elements of `lst`
]]

---


.fatinline[`()`] are used for specifying arguments to functions

- `mean(x)` gives the mean of an array of numbers `x`
- `summary(d)` gives a summary representation of a data frame `d`
- There are several functions used to make the following visualization


```r
ggplot(beaches, aes(x = temperature, y = rainfall)) + 
  geom_point() + 
  geom_smooth() + 
  theme_classic()
```
---


.fatinline[`{}`] are used to contain groups of commands/statements

.pull-left[

A conditional statement


```r
if (age &lt; 18){
   person &lt;- 'Minor'
} else if (age &gt; 65) {
  person &lt;- 'Senior'
} else {
  person &lt;- 'Adult'
}
```
]
.pull-right[

A function definition


```r
my_mean &lt;- function(x, na.rm = T){
  if(na.rm){
    x &lt;- x[!is.na(x)]
  }
  s &lt;- sum(x)
  n &lt;- length(x)
  mn &lt;- s / n # There is a built-in function mean, so I don't use that word
  return(mn)
}
```

]


---

layout: false

---
class: middle,center,inverse

# Packages in R

---



    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script src="../js/macros.js"></script>
<script>var slideshow = remark.create({
"ratio": "16:9",
"highlightLanguage": "r",
"highlightStyle": "tomorrow-night-bright",
"highlightLines": true,
"countIncrementalSlides": false,
"slideNumberFormat": "%current%"
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
